### PR TITLE
UPSTREAM: openshift/source-to-image: 581: fix WaitGroup usage in Execute

### PR DESCRIPTION
**Pull request by gabemontero**
@bparees PTAL

### CODE REVIEW

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
lgtm for post 3.3.  need @deads2k sign off for upstream commit sanity.

File: vendor/github.com/openshift/source-to-image/pkg/build/strategies/sti/sti.go [hunk 1]
```diff
@@ -562,7 +562,6 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 		}()
 
 		opts.Stdin = r
-		defer wg.Wait()
```
Comment (liggitt) [inline vendor/github.com/openshift/source-to-image/pkg/build/strategies/sti/sti.go]: Removing this is definitely correct. Normally there's a single `defer wg.Done()` at the very top of the spawned goroutine

Comment (gabemontero) [inline vendor/github.com/openshift/source-to-image/pkg/build/strategies/sti/sti.go]: @rhcarvalho FYI ^^ from @liggitt re: the re-write you mentioned in https://github.com/openshift/source-to-image/pull/581

Comment (rhcarvalho) [inline vendor/github.com/openshift/source-to-image/pkg/build/strategies/sti/sti.go]: I agree with @liggitt :)

@gabemontero we can write:

```go
go func() {                   // line 544
	defer wg.Done()
	// ...
```

... replacing lines [549](https://github.com/openshift/origin/pull/10776/files#diff-6f7433170556e6fb0dd5276a9f69e422R549) and [559](https://github.com/openshift/origin/pull/10776/files#diff-6f7433170556e6fb0dd5276a9f69e422R559).

Note `wg.Done()` != `wg.Wait()` (the code in this line we're commenting on).

Comment (bparees) [inline vendor/github.com/openshift/source-to-image/pkg/build/strategies/sti/sti.go]: not worth redoing the PR in s2i and then updating this PR for that change though.

we can make the change in s2i and pick it up on the next actual s2i bump if it's deemed that crucial to make the change.  All the return paths are currently covered by the existing logic.

Comment (liggitt) [inline vendor/github.com/openshift/source-to-image/pkg/build/strategies/sti/sti.go]: agree

**Comment by gabemontero** [comment]
bump @bparees or @liggitt for 3.3.1

fyi, the additional change @liggitt noted was dropped into s2i upstream with https://github.com/openshift/source-to-image/commit/563621b87f6aaf711a25ce837bfb74a9b4d2ebe6

**Comment by liggitt** [comment]
1.3.1 or 3.3.1 changes need to be opened against respective branches. Master is now tracking 1.4.0. Not sure if this is worth getting in master or just waiting for a bump

**Comment by gabemontero** [comment]
Ah - makes sense.  I'll redo pull against appropriate branches.  And yeah
there should be at least 1 s2i bump in origin master as part of 1.4.0 so we
can just wait for that.

**Comment by gabemontero** [comment]
closed in lieu of https://github.com/openshift/origin/pull/10968

**Pull request closed without merging**